### DISCUSSION TITLE
Switch to Github Actions

### DISCUSSION THREAD

**Discussion by krainboltgreene** [vcr/vcr#851] [Ideas]
So Travis CI is dead. Anyone wanna create the Github Actions required to test VCR?

**Comment by bradshjg** [comment]
I've been doing this with a few libraries of mine recently, but don't have much experience with Ruby testing. Here's how I see it so far:

- [x] `.travis.yml` -> `.github/workflows/test.yml` keeping the same test matrix and being sensitive to build times (which probably involves caching deps)
- [x] update `README.md` with the new build badge

It's not clear to me whether the main branch test failures are test flakiness or not, so I'll wait until the PR to discuss that part of things.

Edit: looked at this a bit more, and the code folding and timing that `ci.sh` achieves _should_ be possible via GitHub Actions using the [core toolkit grouping](https://github.com/actions/toolkit/blob/main/docs/commands.md#group-and-ungroup-log-lines) but it's not documented whether timing comes for free with that so I'll have to verify.

**Comment by bradshjg** [comment]
Hey @krainboltgreene,

Thanks for switching this to a discussion, I have a prototype of what the move could look like that we can take a look at.

There are a few issues that I'm specifically unsure on:

1. Did I understand that we can probably subsume the s3 caching system into the ruby/ruby-setup actions caching system? AFAICT we're not seeing a ton of deviation in test run performance but I don't know how sensitive the maintainers are to testing performance so I didn't do due diligence in making sure it's running as fast as possible.
2. This is also a bit about performance, but if we attempt to repurpose much of `ci.sh` AFAICT we _can_ keep the grouping/folding but GitHub actions does not add timing for groups (only steps). If timing is important, we could split `ci.sh` out into its constituent steps in the GitHub Actions workflow or do something else I'm not aware of.
3. This probably should have been at the top, but I'm seeing potentially deterministic but at least inconsistent success/failure on runs (it feels like the first run fails and re-runs succeed???). I haven't dug into this much because I wanted to at least make it so we're able to take a look at what things look like in GitHub Actions. One thing I haven't been able to figure out is when I re-run all the jobs in a workflow it appears to hide the initial failure. This has made it difficult to track down the difference between the successful and failed runs against the same commit 😞 

Successful CI run at https://github.com/bradshjg/vcr/actions/runs/608120303

Comparison with vcr default branch at https://github.com/vcr/vcr/compare/master...bradshjg:master

**Reply by krainboltgreene** [reply]
Lets get this merged in ASAP.

**Reply by olleolleolle** [reply]
@bradshjg If you want to make it into a PR, format it the way you want, I can shepherd it into the codebase.

**Reply by bradshjg** [reply]
Sounds good, I'll fix up the badge a bit (at least so that it doesn't reference my fork anymore 😆) but I'm sure there will be some other details to hash out as well. I appreciate the offer to help, I'll need it 😄.
